optimizely/helpers/condition.py

```diff
@@ -233,12 +239,221 @@ def substring_evaluator(self, index):
+
+      Args:
+        target: Given version in string.
+
+      Returns:
+        Boolean:
+            - True if the given version does contain "-"
+            - False if it doesn't
+    """
+        return SemverType.IS_PRE_RELEASE in target
+
+    def is_patch_pre_release(self, idx, idx_value):
+        return idx == SemverType.PATCH_INDEX and idx_value in SemverType.IS_PATCH_PRE_RELEASE
+
+    def is_build(self, target):
+        """ Method to check if the given version contains "+"
+
+      Args:
+        target: Given version in string.
+
+      Returns:
+        Boolean:
+            - True if the given version does contain "+"
+            - False if it doesn't
+    """
+        return SemverType.IS_BUILD in target
+
+    def has_white_space(self, target):
+        """ Method to check if the given version contains " " (white space)
+
+      Args:
+        target: Given version in string.
+
+      Returns:
+        Boolean:
+            - True if the given version does contain " "
+            - False if it doesn't
+    """
+        return SemverType.HAS_WHITE_SPACE in target
+
+    def compare_user_version_with_target_version(self, index):
+        """ Method to compare user version with target version.
+
+    Args:
+      index: Index of the condition to be evaluated.
+
+    Returns:
+      Int:
+        -  0 if user version is equal to target version.
+        -  1 if user version is greater than target version.
+        - -1 if user version is less than target version or, in case of exact string match, doesn't match the target
+        version.
+      None:
+        - if the user version value is not string type or is null.
+    """
+        condition_name = self.condition_data[index][0]
+        target_version = self.condition_data[index][1]
+        user_version = self.attributes.get(condition_name)
+
+        target_version_parts = self.split_semantic_version(target_version)
+        user_version_parts = self.split_semantic_version(user_version)
+        user_version_parts_len = len(user_version_parts)
+
+        for (idx, _) in enumerate(target_version_parts):
+            if user_version_parts_len <= idx:
+                return 1 if self.is_pre_release(target_version) else -1
+            elif not user_version_parts[idx].isdigit():
+                if user_version_parts[idx] < target_version_parts[idx]:
+                    return -1
+                elif user_version_parts[idx] > target_version_parts[idx]:
+                    return 1
+            else:
+                user_version_part = int(user_version_parts[idx])
+                target_version_part = int(target_version_parts[idx])
+                if user_version_part > target_version_part:
+                    return 1
+                elif user_version_part < target_version_part:
+                    return -1
+        if self.is_pre_release(user_version) and not self.is_pre_release(target_version):
+            return -1
+        return 0
```

Comment:
nit. It will be nice to have these defined before they are referenced.

Reply:
done